packages/hub/services/gas-estimation.ts

```diff
@@ -39,24 +39,71 @@ export default class GasEstimationService {
     let provider = this.ethersProvider.getInstance(params.chainId);
     let signer = new Wallet(config.get('hubPrivateKey'));
     let scheduledPaymentModule = await this.cardpay.getSDK('ScheduledPaymentModule', provider, signer);
+
     let gas;
-    let tokenAmount = ethers.utils.parseUnits('1', 'gwei');
-    let gasTokenAmount = ethers.utils.parseUnits('0.01', 'gwei');
     switch (params.scenario) {
       case GasEstimationResultsScenarioEnum.create_safe_with_module:
         gas = (
           await scheduledPaymentModule.createSafeWithModuleAndGuardEstimation({ from: signer.address })
         ).toNumber();
         break;
+      default:
+        gas = await this.estimatePaymentExecution(params);
+    }
+
+    gasLimit = await prisma.gasEstimationResult.upsert({
+      where: {
+        chainId_scenario_tokenAddress_gasTokenAddress: {
+          chainId: params.chainId,
+          scenario: params.scenario,
+          tokenAddress: params.tokenAddress ?? '',
+          gasTokenAddress: params.gasTokenAddress ?? '',
+        },
+      },
+      create: {
+        chainId: params.chainId,
+        scenario: params.scenario,
+        tokenAddress: params.tokenAddress,
+        gasTokenAddress: params.gasTokenAddress,
+        gas: gas,
+      },
+      update: {
+        gas: gas,
+      },
+    });
+
+    return gasLimit;
+  }
+
+  private async estimatePaymentExecution(params: GasEstimationParams) {
+    if (
+      !params.tokenAddress ||
+      params.tokenAddress === '' ||
+      !params.gasTokenAddress ||
+      params.gasTokenAddress === ''
+    ) {
+      throw Error(`tokenAddress and gasTokenAddress is required in ${params.scenario}`);
+    }
+
+    let provider = this.ethersProvider.getInstance(params.chainId);
+    let networkName = convertChainIdToName(params.chainId);
+    let tokenList = this.cardpay.getConstantByNetwork('tokenList', networkName as SchedulerCapableNetworks);
+    let token = tokenList.tokens.find((t) => t.address === params.tokenAddress);
+    let gasToken = tokenList.tokens.find((t) => t.address === params.gasTokenAddress);
+    if (!token || !gasToken) {
+      throw Error('unknown token and gas token');
+    }
+    //Check the decimals of tokens
+    //to avoid error when converting USD tokens
+    //because USD token's decimals is lower than native token's decimals
+    let tokenAmount = ethers.utils.parseUnits('1', token.decimals >= 18 ? 'gwei' : 3);
```

Comment:
Can you explain why 3? Can we simplify this, would this work? 🔽

```
let tokenAmount = ethers.utils.parseUnits('1', token.decimals);
let gasTokenAmount = ethers.utils.parseUnits('0.01', token.decimals);
```

Reply:
Updated like this:

`let tokenAmount = ethers.utils.parseUnits('1', token.decimals >= 18 ? 'gwei' : token.decimals);
let gasTokenAmount = ethers.utils.parseUnits('0.01', gasToken.decimals >= 18 ? 'gwei' : token.decimals);`

If we use `ethers.utils.parseUnits('1', token.decimals);` it will be too big to have 1 eth for the crank safe.